iscsi/targetd/cmd/start.go

```diff
@@ -0,0 +1,135 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"github.com/Sirupsen/logrus"
+	"github.com/kubernetes-incubator/external-storage/iscsi/targetd/provisioner"
+	"github.com/kubernetes-incubator/external-storage/lib/controller"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"time"
+)
+
+var log = logrus.New()
+
+// start-controllerCmd represents the start-controller command
+var startcontrollerCmd = &cobra.Command{
+	Use:   "start",
+	Short: "A brief description of your command",
+	Long: `A longer description that spans multiple lines and likely contains examples
+and usage of using your command. For example:
+
+Cobra is a CLI library for Go that empowers applications.
+This application is a tool to generate the needed files
+to quickly create a Cobra application.`,
+	Run: func(cmd *cobra.Command, args []string) {
+		initLog()
+
+		log.Debugln("start called")
+
+		// creates the in-cluster config
+		log.Debugln("creating in cluster default kube client config")
+		config, err := rest.InClusterConfig()
+		if err != nil {
+			log.Fatalln(err)
+		}
+		log.WithFields(logrus.Fields{
+			"config-host": config.Host,
+		}).Debugln("kube client config created")
+
+		//	 creates the clientset
+		log.Debugln("creating kube client set")
+		kubernetesClientSet, err := kubernetes.NewForConfig(config)
+		if err != nil {
+			log.Fatalln(err)
+		}
+		log.Debugln("kube client set created")
+
+		// The controller needs to know what the server version is because out-of-tree
+		// provisioners aren't officially supported until 1.5
+		serverVersion, err := kubernetesClientSet.Discovery().ServerVersion()
+		if err != nil {
+			log.Fatalf("Error getting server version: %v", err)
+		}
+
+		//url := viper.GetString("targetd-scheme") + "://" + viper.GetString("targetd-username") + ":" + viper.GetString("targetd-password") + "@" + viper.GetString("targetd-address") + ":" + viper.GetInt("targetd-port")
+
+		url := fmt.Sprintf("%s://%s:%s@%s:%d/targetrpc", viper.GetString("targetd-scheme"), viper.GetString("targetd-username"), viper.GetString("targetd-password"), viper.GetString("targetd-address"), viper.GetInt("targetd-port"))
+
+		log.Debugln("targed URL", url)
+
+		iscsiProvisioner := provisioner.NewiscsiProvisioner(url)
+		log.Debugln("iscsi provisioner created")
+		pc := controller.NewProvisionController(kubernetesClientSet, viper.GetDuration("resync-period"), viper.GetString("provisioner-name"), iscsiProvisioner, serverVersion.GitVersion,
```

Comment:
For every argument to NewProvisionController you have to modify it like: replace ```viper.GetDuration("resync-period")``` with ```controller.ResyncPeriod(viper.GetDuration("resync-period")```.  Order doesn't matter anymore except for the first 4 arguments.

I opened #190 because you may want to replace the flag defaults too or you may want to choose your own e.g. ```startcontrollerCmd.Flags().Duration("resync-period", 15*time.Second, "how often to poll the master API for updates")``` with ```startcontrollerCmd.Flags().Duration("resync-period", DefaultResyncPeriod, "how often to poll the master API for updates")```

The reason for the new different syntax is I didn't anticipate most provisioners to need to modify the defaults so as to not burden people with having to learn what every parameter means. (Like here we just use the defaults: https://github.com/kubernetes-incubator/external-storage/blob/master/aws/efs/cmd/efs-provisioner/efs-provisioner.go#L264). But exposing them via optional flags is a good idea like done here.